mlflow/crewai/autolog.py

```diff
@@ -0,0 +1,228 @@
+import inspect
+import json
+
+import mlflow
+from mlflow.entities import SpanType
+from mlflow.entities.span import LiveSpan
+from mlflow.tracing.utils import TraceJSONEncoder
+from mlflow.utils.autologging_utils.config import AutoLoggingConfig
+
+
+def patched_class_call(original, self, *args, **kwargs):
+    config = AutoLoggingConfig.init(flavor_name=mlflow.gemini.FLAVOR_NAME)
+
+    if config.log_traces:
+        fullname = f"{self.__class__.__name__}.{original.__name__}"
+        with mlflow.start_span(
+            name=fullname,
+            span_type=_get_span_type(self.__class__, fullname),
+        ) as span:
+            inputs = _construct_full_inputs(original, self, *args, **kwargs)
+            span.set_inputs(inputs)
+            _set_span_attributes(span=span, instance=self)
+            result = original(self, *args, **kwargs)
+            # need to convert the response of generate_content for better visualization
+            outputs = result.__dict__ if hasattr(result, "__dict__") else result
+            span.set_outputs(outputs)
+
+            return result
+
+
+def _get_span_type(cls, task_name: str) -> str:
+    from crewai.flow.flow import Flow
+
+    if issubclass(cls, Flow):
+        return SpanType.CHAIN
+    span_type_mapping = {
+        "Crew.kickoff": SpanType.CHAIN,
+        "Crew.kickoff_for_each": SpanType.CHAIN,
+        "Crew.train": SpanType.CHAIN,
+        "Agent.execute_task": SpanType.AGENT,
+        "Task.execute_sync": SpanType.CHAIN,
+        "ShortTermMemory.save": SpanType.RETRIEVER,
+        "ShortTermMemory.search": SpanType.RETRIEVER,
+        "ShortTermMemory.reset": SpanType.RETRIEVER,
+        "LongTermMemory.save": SpanType.RETRIEVER,
+        "LongTermMemory.search": SpanType.RETRIEVER,
+        "LongTermMemory.reset": SpanType.RETRIEVER,
+        "UserMemory.search": SpanType.RETRIEVER,
+        "UserMemory.save": SpanType.RETRIEVER,
+        "EntityMemory.search": SpanType.RETRIEVER,
+        "EntityMemory.save": SpanType.RETRIEVER,
+        "EntityMemory.reset": SpanType.RETRIEVER,
+        "Knowledge.query": SpanType.RETRIEVER,
+        "LLM.call": SpanType.LLM,
+        "Flow.kickoff": SpanType.CHAIN,
+    }
+    return span_type_mapping.get(task_name, SpanType.UNKNOWN)
+
+
+def _is_serializable(value):
+    try:
+        json.dumps(value, cls=TraceJSONEncoder, ensure_ascii=False)
+        return True
+    except (TypeError, ValueError):
+        return False
```

Comment:
Is there any way we can check this (even in exchange to some cost of accuracy) without actually serializing the object? For context, the serialization actually has visible overhead right now for large spans e.g. document retrieval.

Reply:
We could check acceptable classes of the encoder first and then also check the references of the pydamic model. But if you care the latency due to a large responce from llm, I didn't call this logic for outputs.
https://docs.pydantic.dev/latest/concepts/forward_annotations/#cyclic-references

Reply:
> But if you care the latency due to a large responce from llm, I didn't call this logic for outputs.

Interesting. Would you elaborate why we only need this for input and what happens if we remove this?

Reply:
Btw I faced this warning when running the code on my notebook. Is this perhaps related to this logic?
```
2024/12/02 01:23:52 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/databricks/python/lib/python3.12/site-packages/pydantic/functional_validators.py:659: UserWarning: Pydantic serializer warnings:
  Expected `InstanceOf` but got `CrewAgentExecutor` - serialized value may not be as expected"
```

Reply:
The input object has some attributes that have cyclic references, and the serialization fails because of the cyclic references. We could add a validation if the class is defined by us like the code in the pydantic documentation, but since this class is defined by crewai library, the safest way is trying to serialize, like my code here. 
https://docs.pydantic.dev/latest/concepts/forward_annotations/#self-referencing-or-recursive-models

Reply:
The warning occurs due to the mismatch between the expected class and the actual object inside the crewai.Agent implementation. It is displayed whenever the agent object is serialized.